extensions/amp-ad/0.1/amp-ad-3p-impl.js

```diff
@@ -0,0 +1,438 @@
+    // have changed. Send an intersection record if needed. This does nothing
+    // if we aren't currently in view.
+    if (this.intersectionObserver_) {
+      this.intersectionObserver_.fire();
+    }
+  }
+
+  /**
+   * Measure the layout box of the iframe if we rendered it already.
+   * @private
+   */
+  measureIframeLayoutBox_() {
+    if (this.iframe_) {
+      this.iframeLayoutBox_ =
+          this.getViewport().getLayoutRect(this.iframe_);
+    }
+  }
+
+  /**
+   * @override
+   */
+  getIntersectionElementLayoutBox() {
+    if (!this.iframe_) {
+      return super.getIntersectionElementLayoutBox();
+    }
+    if (!this.iframeLayoutBox_) {
+      this.measureIframeLayoutBox_();
+    }
+    return this.iframeLayoutBox_;
+  }
+
+  /** @override */
+  layoutCallback() {
+    if (!this.iframe_) {
+      user.assert(!this.isInFixedContainer_,
+          '<amp-ad> is not allowed to be placed in elements with ' +
+          'position:fixed: %s', this.element);
+      incrementLoadingAds(this.getWin());
+      return getAdCid(this).then(cid => {
+        if (cid) {
+          this.element.setAttribute('ampcid', cid);
+        }
+        this.iframe_ = getIframe(this.element.ownerDocument.defaultView,
+            this.element);
+        this.iframe_.setAttribute('scrolling', 'no');
+        this.applyFillContent(this.iframe_);
+        this.intersectionObserver_ =
+            new IntersectionObserver(this, this.iframe_, /* opt_is3P */true);
+        // Triggered by context.noContentAvailable() inside the ad iframe.
+        listenForOnce(this.iframe_, 'no-content', () => {
+          this.noContentHandler_();
+        }, /* opt_is3P */ true);
+        // Triggered by context.reportRenderedEntityIdentifier(…) inside the ad
+        // iframe.
+        listenForOnce(this.iframe_, 'entity-id', info => {
+          this.element.creativeId = info.id;
+        }, /* opt_is3P */ true);
+        listenFor(this.iframe_, 'embed-size', data => {
+          let newHeight, newWidth;
+          if (data.width !== undefined) {
+            newWidth = Math.max(this.element./*OK*/offsetWidth +
+                data.width - this.iframe_./*OK*/offsetWidth, data.width);
+            this.iframe_.width = data.width;
+            this.element.setAttribute('width', newWidth);
+          }
+          if (data.height !== undefined) {
+            newHeight = Math.max(this.element./*OK*/offsetHeight +
+                data.height - this.iframe_./*OK*/offsetHeight, data.height);
+            this.iframe_.height = data.height;
+            this.element.setAttribute('height', newHeight);
+          }
+          if (newHeight !== undefined || newWidth !== undefined) {
+            this.updateSize_(newHeight, newWidth);
+          }
+        }, /* opt_is3P */ true);
+        this.iframe_.style.visibility = 'hidden';
+        listenForOnce(this.iframe_, 'render-start', () => {
+          this.iframe_.style.visibility = '';
```

Comment:
This will need to be synced with https://github.com/ampproject/amphtml/pull/3670